core/src/main/scala/kafka/message/MessageAndOffset.scala

```diff
@@ -17,11 +17,31 @@
 
 package kafka.message
 
-import org.apache.kafka.common.record.AbstractLegacyRecordBatch
+import org.apache.kafka.common.record.{AbstractLegacyRecordBatch, Record, RecordBatch}
 
 object MessageAndOffset {
-  def fromRecordBatch(recordBatch: AbstractLegacyRecordBatch): MessageAndOffset = {
-    MessageAndOffset(Message.fromRecord(recordBatch.outerRecord), recordBatch.lastOffset)
+  def fromRecordBatch(batch: RecordBatch): MessageAndOffset = {
+    batch match {
+      case legacyBatch: AbstractLegacyRecordBatch =>
+        val legacyRecordBatch = batch.asInstanceOf[AbstractLegacyRecordBatch]
+        MessageAndOffset(Message.fromRecord(legacyRecordBatch.outerRecord), legacyRecordBatch.lastOffset)
+
+      case _ =>
+        throw new IllegalArgumentException(s"Illegal batch type ${batch.getClass}. The older message format classes " +
+          s"only support conversion from ${classOf[AbstractLegacyRecordBatch]}, which is used for magic v0 and v1")
+    }
+  }
+
+  def fromRecord(record: Record): MessageAndOffset = {
+    record match {
+      case legacyBatch: AbstractLegacyRecordBatch =>
```

Comment:
It's so weird how the record and record batch are of the same type. Happy that we're moving away from that.